Library/Homebrew/livecheck/strategy/github_latest.rb

```diff
@@ -0,0 +1,73 @@
+# typed: false
+# frozen_string_literal: true
+
+module Homebrew
+  module Livecheck
+    module Strategy
+      # The {GithubLatest} strategy identifies versions of software at
+      # github.com by checking a repository's "latest" release page.
+      #
+      # GitHub URLs take a few different formats:
+      #
+      # * `https://github.com/example/example/releases/download/1.2.3/example-1.2.3.tar.gz`
+      # * `https://github.com/example/example/archive/v1.2.3.tar.gz`
+      # * `https://github.com/downloads/example/example/example-1.2.3.tar.gz`
+      #
+      # A repository's `/releases/latest` URL normally redirects to a release
+      # tag (e.g., `/releases/tag/1.2.3`). When there isn't a "latest" release,
+      # it will redirect to the `/releases` page.
+      #
+      # This strategy should only be used when we know the upstream repository
+      # has a "latest" release and the tagged release is appropriate to use
+      # (e.g., "latest" isn't wrongly pointing to an unstable version, not
+      # picking up the actual latest version, etc.). The strategy can only be
+      # applied by using `strategy :github_latest` in a `livecheck` block.
+      #
+      # The default regex identifies versions like `1.2.3`/`v1.2.3` in `href`
+      # attributes containing the tag URL (e.g.,
+      # `/example/example/releases/tag/v1.2.3`). This is a common tag format
+      # but a modified regex can be provided in a `livecheck` block to override
+      # the default if a repository uses a different format (e.g.,
+      # `example-1.2.3`, `1.2.3d`, `1.2.3-4`, etc.).
+      #
+      # @api public
+      class GithubLatest
+        NICE_NAME = "GitHub - Latest"
+
+        # A priority of zero causes livecheck to skip the strategy. We do this
+        # for {GithubLatest} so we can selectively apply the strategy using
+        # `strategy :github_latest` in a `livecheck` block.
+        PRIORITY = 0
```

Comment:
Previously, 0 priority was equivalent (specific) to the PageMatch strategy.
Now, we introduce another strategy with the same priority. This makes the way we write conditionals that process these priorities important.

Reply:
The original idea was that `PRIORITY = 0` would cause a strategy to be omitted unless it's specified in a `livecheck` block (e.g., `strategy :github_latest`). I think the updated logic in `Livecheck::Strategy#from_url` more accurately represents this, as it now ensures the strategy can be applied when specified in a `livecheck` block.

Since `PageMatch` was the only strategy with a priority of zero up to this point and it was special-cased, I forgot to account for the `from_symbol(livecheck_strategy) != strategy` condition originally. Outside of this omission, we haven't treated `PRIORITY = 0` as being equivalent to `strategy == PageMatch` in livecheck's code and we specifically reference `PageMatch` (or `:page_match`) where appropriate.

`PageMatch` is an exception to the above idea, rather than the rule. The `PageMatch` strategy will automatically apply if a `livecheck` block contains a `url` and `regex`, so it's not omitted like other zero-priority strategies. The reasons why it has a priority of zero is to be able to enforce the regex requirement before applying it and to also deprioritize it with respect to more specific strategies (e.g., those with the `DEFAULT_PRIORITY` of 5 or a higher `PRIORITY`).

---

I've thought about updating the `#match?` method in strategies to allow for contextual information to be passed in. I imagine this would either be the `Livecheck` object (i.e., the information in the formula's `livecheck` block) or maybe an optional `options` hash (to allow for more flexibility/extensibility in the future). With this setup, we could give `PageMatch` a priority of 1 (deprioritized but not omitted) and enforce the `regex` requirement in `PageMatch#match?`.

I believe this is a better setup overall, as we wouldn't violate the `PRIORITY = 0` omission idea above and also it would allow us to keep this information in strategies instead of `Strategy#from_url`. I'll create a PR for this after I've finished the livecheck docs PR (and I allow myself to work on livecheck's internals again) and we can discuss it further there.

---

That said, I believe the related changes in this PR are sufficient for now but let me know if I've overlooked anything important.